### PR TITLE
WebGLRenderer: Make sure stencil works with transmission.

### PR DESCRIPTION
**Pull request by Mugen87**
Related #26959.

**Description**

This PR is an alternative implementation to #26959.

When the default drawing buffer has a stencil buffer, this PR makes sure the transmission render target has one as well. This is achieved by applying the `stencil` context parameter to the internal render target.

This PR is more simple but less flexible than #26959 since it does not perform dynamic checks on render targets. The policy is if stencil is required in the transmission pass, is has to be turned on for the default drawing buffer.

### CODE REVIEW

**Comment by mrdoob** [comment]
@Oletus Solves the issue for you?

**Comment by Mugen87** [comment]
Syncing stencil of the transmission render target with the default framebuffer seems a step in the right direction. We can discuss in https://github.com/mrdoob/three.js/pull/26959 if a more dynamic solution is still required.

**Comment by Mugen87** [comment]
@mrdoob It has been reported in https://github.com/mrdoob/three.js/issues/27846#issuecomment-1985524127 that enabling stencil for transmission has a negative impact on the performance.

I still think that syncing the stencil buffer of the transmission render target with the setup of the drawing buffer is right. But that leads me to the question: Why is the `stencil` context attribute `true` by default? Couldn't we set it to `false` for performance reasons? Only when stencil is required by an application, it is set to `true`.

**Comment by lo-th** [comment]
here the version with stencil on
https://lo-th.github.io/phy/index2.html#random

**Comment by Mugen87** [comment]
I get 25 - 55 with stencil and 26-55 without on a mac Mini with M2 Pro. But I admit a performance degradation is visible.

**Comment by lo-th** [comment]
26 !? wow mac :)
i have strong 144 - 60 normally and 25 - 25 with stencil on

**Comment by Mugen87** [comment]
This kind of FPS difference should not be the case. An additional stencil buffer can't be the sole reason this kind of performance degradation. This is almost worth reporting at the browser vendor. What kind of browser and operating system are you using?

**Comment by lo-th** [comment]
windows 11 / chrome / rtx 4090  / threadripper 32 core :)
is same with firefox ( firefox have lower perf on physx worker )

**Comment by mrdoob** [comment]
> But that leads me to the question: Why is the `stencil` context attribute `true` by default? Couldn't we set it to `false` for performance reasons? Only when stencil is required by an application, it is manually set to `true`.

I don't think there was a reason for it.
Setting it to false by default sounds good to me 👌

**Comment by Mugen87** [comment]
@lo-th I've researched a bit more and it seems the following document explains the performance drop on Windows:

https://groups.google.com/g/webgl-dev-list/c/iwIYWPPJwoQ?pli=1

tl;dr: D3D does not support resolving depth+stencil on the GPU. Hence, WebGL must do this on the CPU (emulation) which is slow. Since the transmission render target is multisampled, it goes through this process on Windows when `stencilBuffer` is set to `true` which explains the poor performance.

**Comment by Oletus** [comment]
I think there should be some way of disabling the stencil bit when resolving using `blitFramebuffer`, even if the multisample rendertarget uses a stencil buffer. There are applications which use stencil while rendering, but only care about color buffer after resolving. Maybe a `resolveStencil` bit in `WebGLRenderTarget` that would be `false` by default?

**Comment by Mugen87** [comment]
TBH, I would prefer a solution that doesn't require a new flag. At least not a flag that has to be set by users.

> There are applications which use stencil while rendering, but only care about color buffer after resolving

Maybe we can always do this for the transmission render target and nobody will notice^^?

**Comment by Oletus** [comment]
Yep, avoiding a flag is good if we can get away with it! I don't think there will be any problems with resolving just the color buffer in case of the transmission render target. But it's a bit harder to say in general what's the most appropriate for user-created multisample render targets.

**Comment by Mugen87** [comment]
I have implemented the discussed approach in https://github.com/Mugen87/three.js/commit/707d4548626e176242a62925bf0c167e88b061e9 for testing:

1. Transmission with stencil (but resolve): https://rawcdn.githack.com/Mugen87/three.js/6534c98c7b603a1b19408763b3c372f17a0dceae/examples/webgl_loader_gltf_transmission.html?stenci=true
2. Transmission with stencil (with resolve):
https://rawcdn.githack.com/mrdoob/three.js/d3c9871672acc7b38120261513eeccc0dc862640/examples/webgl_loader_gltf_transmission.html
3. Transmission without stencil: https://rawcdn.githack.com/Mugen87/three.js/6534c98c7b603a1b19408763b3c372f17a0dceae/examples/webgl_loader_gltf_transmission.html?stencil=false

@lo-th Do you still see a performance degradation in the first link on Windows? Ideally test 1. and 3. should have equal performance, test 2. should show a performance drop (it's the version before the revert https://github.com/mrdoob/three.js/pull/27885).

**Comment by lo-th** [comment]
i see no difference on both ?
in my demo i also use mirror on floor

**Comment by Mugen87** [comment]
Maybe you can apply the patch to your demo and check the performance? Update build files are available here: 

https://github.com/Mugen87/three.js/commit/6534c98c7b603a1b19408763b3c372f17a0dceae

**Comment by lo-th** [comment]
work fine with this build and stencil=true

**Comment by Oletus** [comment]
@Mugen87 Hey, it looks like stencil buffer bit is still set when resolving MSAA render targets, unless it's a transmission render target? Any chance we could universally disable stencil multisample resolve, or implement a flag controlling that?

**Comment by Mugen87** [comment]
> Hey, it looks like stencil buffer bit is still set when resolving MSAA render targets, unless it's a transmission render target?

Correct. The following line controls this:

https://github.com/mrdoob/three.js/blob/c3b2fb018a8b4872b64013f36c989dfaba43b9d0/src/renderers/webgl/WebGLTextures.js#L1881-L1883

I'm open for a `resolveStencil` flag in the `RenderTarget` class. The internal transmission render target could set it to `false` so we would not need `__isTransmissionRenderTarget` anymore.
